src/traces/heatmap/hover.js

```diff
@@ -120,8 +106,26 @@ module.exports = function hoverPoints(pointData, xval, yval, hovermode, hoverLay
         y1: y1,
         xLabelVal: xl,
         yLabelVal: yl,
-        zLabelVal: zVal,
-        zLabel: zLabel,
         text: text
-    })];
+    };
+
+    var zVal = z[ny][nx];
+    if(zmask && !zmask[ny][nx]) zVal = undefined;
+
+    if(zVal !== undefined || trace.hovergaps) {
+        // dummy axis for formatting the z value
+        var cOpts = extractOpts(trace);
+        var dummyAx = {
+            type: 'linear',
+            range: [cOpts.min, cOpts.max],
+            hoverformat: zhoverformat,
+            _separators: xa._separators,
+            _numFormat: xa._numFormat
+        };
+
+        obj.zLabelVal = zVal;
+        obj.zLabel = Axes.tickText(dummyAx, zVal, 'hover').text;
+    }
+
+    return [Lib.extendFlat(pointData, obj)];
```

Comment:
So this mean when `hovergaps:false`, `zLabel` and `zLabelVal` remain undefined, but we still pass `pointData` back to `Fx.hover`?

Does this mean that the hover label is suppressed, but the `plotly_hover` event is still getting triggered?

Reply:
Here is a [codepen](https://codepen.io/MojtabaSamimi/pen/yLLgoNN?editors=0010) displaying the current behaviour with `hoverongaps` disabled.
Please let me know if you prefer not to have hover effects at all.
cc: @emmanuelle

Reply:
I think the behaviour in @archmoj above [codepen](https://codepen.io/MojtabaSamimi/pen/yLLgoNN?editors=0010) is incorrect. Under `hoverongaps: false`, there should no hover labels shown and no `plotly_hover` event triggered.

Reply:
Now fixed in https://github.com/plotly/plotly.js/pull/4291/commits/92693525664f8d067385b410f268d2b36b3f058e.
[new demo](https://codepen.io/MojtabaSamimi/pen/WNNRqdg?editors=0010)

Reply:
Fantastic!!